### DISCUSSION TITLE
Enable passing HTTP headers in data query requests to data source backend endpoints

### DISCUSSION THREAD

**Discussion by zhichli** [grafana/grafana#58287] [Data Sources]
Currently, for requests of `api/ds/query`, the [query service](https://github.com/grafana/grafana/blob/a83fdc6b879127d506f0ffc3639a25cdf36bf246/pkg/services/query/query.go#L162) is not passing request HTTP headers to data source backend, which seems intentional for data query requests based on [this comment](https://github.com/grafana/grafana/pull/51061#discussion_r904927726) since the `headers` field of [`QueryDataRequest`](https://github.com/grafana/grafana-plugin-sdk-go/blob/aab437cf0c85570d06c05e6b43e63df967329bf0/proto/backend.proto#L117-L118) is designed for internally use only on environment info. But there is some use scenario that **passing request HTTP headers to data source backend endpoints** could be useful from the perspective of business insights and supportability. 

I admit this might be a bit special use scenario but it is mainly for demonstrating the motivation for this feature request.

**Scenario:** 
We runs a managed grafana service (suppose service name is ABC) and have a reverse proxy (e.g. ABCProxy) running in front of Grafana, and the proxy injects a HTTP Header `User-Agent` with value of the proxy's build version e.g. `User-Agent: ABC/build-1024`. And we expect this header to be passed by Grafana to data source endpoint for all kinds of requests (it already works for resources requests `api/datasources/<id>/resources/*` but **not** for data query requests `api/ds/query`). And the data source backend endpoint is owned by our partner team running a managed Prometheus service with endpoint like "https://managed.prometheus.com/". Then the service owner of the managed Prometheus team can log the `User-Agent` header value in their service logs, which can be used for multiple purposes:
- telling if the request is coming from "ABC" managed grafana rather than self-hosted grafana (self-hosted Grafana will [set "User-Agent" header with value of "Grafana/<version>" (e.g. "Grafana/9.2.0")](https://github.com/grafana/grafana/blob/a83fdc6b879127d506f0ffc3639a25cdf36bf246/pkg/infra/httpclient/httpclientprovider/user_agent_middleware.go#L22). This way, the service owner of Azure Managed Prometheus team can gain business insights of AMG's requests versus self-hosted Grafana's requests. 
- telling the exact build version of the downstream ABC service, so they can provide this info when troubleshooting incidents together with ABC team
- they might implement different throttling or load balancing strategy based on the header value

The motivations are similar to this [feature ask](https://github.com/grafana/grafana/issues/54403#issuecomment-1248235467)

To make this feature work, there seems to be at least 2 changes needed:
- copy [headers of `parsedReq.httpRequest`](https://github.com/grafana/grafana/blob/a83fdc6b879127d506f0ffc3639a25cdf36bf246/pkg/services/query/query.go#L211) into [the `Headers` field of QueryDataRequest](https://github.com/grafana/grafana/blob/a83fdc6b879127d506f0ffc3639a25cdf36bf246/pkg/services/query/query.go#L188) 
- add requests headers to [context here](https://github.com/grafana/grafana/blob/bff19747bdbd07c511b5351c66b110a6f8edc129/pkg/tsdb/prometheus/buffered/time_series_query.go#L108)
 
We wonder if this feature can be reconsidered.

**Comment by wbrowne** [comment]
@grafana/plugins-platform-backend

**Reply by marefr** [reply]
As I wrote in https://github.com/grafana/grafana/pull/58094#discussion_r1031585717:
> First of all these changes is blocked by https://github.com/grafana/grafana/pull/58132, which refactors everything related to forwarding of HTTP headers, such as Cookies and OAuth token, and is targeted for Grafana v9.4.0.
>
> Second, we don't want to do it like this because `req.Headers` is not meant for solely HTTP headers - it's purpose is environment info/custom metadata, see
https://github.com/grafana/grafana-plugin-sdk-go/blob/701ffb57ba2dd097fc765991cba38100ea41a41f/proto/backend.proto#L122-L123
https://github.com/grafana/grafana-plugin-sdk-go/blob/701ffb57ba2dd097fc765991cba38100ea41a41f/proto/backend.proto#L175-L176
>
> Third, forwarding any HTTP header might have security implications so we might need to consider an allow/deny list of headers allowed to be forwarded.
>
> We need to figure out how to allow HTTP headers to be set in this environment info. Here's a WIP/proposal of what this might look like: https://github.com/grafana/grafana-plugin-sdk-go/pull/562. I suggest that we continue this discussion in https://github.com/grafana/grafana/discussions/58287

So https://github.com/grafana/grafana-plugin-sdk-go/pull/562 introduces a new interface that should make it easier for plugin authors to retrieve HTTP headers using the same interface for `QueryDataRequest`, `CheckHealthRequest` and `CallResourceRequest`.

For Grafana this interface could also be use to forward HTTP headers in plugin requests, i.e.  `QueryDataRequest`, `CheckHealthRequest` and `CallResourceRequest`.

What needs to be discussed is if above interface is a good idea or not and the potential security implications, i.e.
- Need header allow/deny list configuration on Grafana level to decide what headers being forwarded if available in incoming HTTP request?
- Need header allow configuration on each HTTP datasource configuration, similar to existing `Allowed cookies` settings to decide on a per datasource what headers being forwarded if available in incoming HTTP request?
![image](https://user-images.githubusercontent.com/1668778/203812939-38e2caff-4816-4495-8a9b-fff18e38b214.png)
- Any incoming HTTP header, e.g. API key, auth proxy/JWT headers, being used for authenticating a user would never be forwarded to a plugin. It's enforced by recent security changes introduced in https://github.com/grafana/grafana/pull/57399 and outlined in the [Grafana security releases: New versions with fixes for CVE-2022-39229, CVE-2022-39201, CVE-2022-31130, CVE-2022-31123](https://grafana.com/blog/2022/10/12/grafana-security-releases-new-versions-with-fixes-for-cve-2022-39229-cve-2022-39201-cve-2022-31130-cve-2022-31123/) blog post

Thoughts? Let me know if something is unclear and I'll try to make it more clear.

**Reply by marefr** [reply]
@zhichli in regards to https://github.com/grafana/grafana/discussions/58714 how do you imagine this to work since Grafana will always set it's `User-Agent` for outgoing HTTP requests. Seems to me this is either a duplicate of the other one or they are two different asks for the same thing?

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
